src/storage-controller/src/collection_mgmt.rs

```diff
@@ -192,13 +346,377 @@ where
     }
 }
 
+/// Spawns an [`mz_ore::task`] that will make it so that the state in persist
+/// matches the desired state and continuously bump the upper for the specified
+/// collection.
+///
+/// NOTE: This implementation is a bit clunky, and could be optimized by not keeping
+/// all of desired in memory (see commend below). It is meant to showcase the
+/// general approach.
```

Comment:
I think for unified compute introspection, doing this optimization would be pretty important. When discussing this feature in compute, we explicitly rules out keeping all the introspection data in memory to be able to retract it later, to avoid having envd memory usage scale with the number of replicas. It may be fine as long as we only have a few unified introspection collections, but I'd feel better to have this optimization in the initial implementation so we are not tempted to punt on it later.

Reply:
Discussed this offline: For the initial version of unified introspection the current way is fine. It actually simplifies the deletion of entries when a replica is dropped since we can just delete them from the in memory `desired` state, instead of having to read back the shard. As we start adding unified introspection collections we should start thinking about implementing the optimization.